stdlib/regex.gr

```diff
@@ -0,0 +1,2983 @@
+          } else {
+            let rec dLoop = (pos, accum) => {
+              if (pos == len) {
+                [getInputSubstr(accum)]
+              } else {
+                let c = replacementExploded[pos]
+                if (Char.code('0') <= Char.code(c) && Char.code(c) <= Char.code('9')) {
+                  dLoop(pos + 1, (10 * accum) + (Char.code(c) - Char.code('0')))
+                } else {
+                  [getInputSubstr(accum), ...loop(pos, pos)]
+                }
+              }
+            }
+            dLoop(pos + 1, 0)
+          }
+        })
+      }
+    } else {
+      loop(pos + 1, since)
+    }
+  }
+  let res = loop(0, 0)
+  List.reduceRight(String.concat, "", res)
+}
+
+
+let regexReplaceHelp = (rx: RegularExpression, toSearch: String, replacement: String, all: Bool) => {
+  let buf = makeMatchBuffer(toSearch)
+  let mut out = []
+  let rec loop = (searchPos) => {
+    let state = Array.make(rx.reNumGroups, None)
+    let poss = searchMatch(rx, buf, searchPos, searchPos, Array.length(buf.matchInputExploded), state)
+    let recur = (start, end) => {
+      if (end == searchPos) {
+        if (searchPos == String.length(toSearch)) {
+          ""
+        } else {
+          String.concat(String.slice(searchPos, searchPos + 1, toSearch), loop(searchPos + 1))
+        }
+      } else {
+        loop(end)
+      }
+    }
+    match(poss) {
+      None => if (searchPos == 0) toSearch else String.slice(searchPos, String.length(toSearch), toSearch),
+      Some((start, end)) =>
+        String.concat(String.slice(searchPos, start, toSearch),
+          String.concat(computeReplacement(buf, replacement, start, end, state),
+            if (all) recur(start, end) else String.slice(end, String.length(toSearch), toSearch))),
+    }
+  }
+  loop(0)
+}
+
+/**
+ * Replaces the first match for the given regular expression contained within the given string with the given replacement.
+ * Replacement strings support the following syntax:
+ * - `$&` - Replaced with the text of the matching portion of input (e.g. for `(foo)`, the search string `foo bar`, and the replacement `baz $&`, the result will be `baz foo bar`)
+ * - `$n` / `$nn` (where `n` is a digit) - Replaced with the text of group `nn`
+ * - `$$` - Replaced with a literal `$`
+ * - `$.` - Does nothing (this exists to support replacement strings such as `$4$.0`, which will place the contents of group 4 prior to a zero)
+ * - `$\`` - Replaced with the text of the string prior to the matched subspan of text
+ * - `$'` - Replaced with the text of the string after the matched subspan of text
+ * - Any other character will be placed as-is in the replaced output.
+ *
+ * @param rx: RegularExpression - The regular expression to search for
+ * @param toSearch: String - The string to search within
+ * @param replacement: String - The string to replace matches with
+ * @returns The `replacement` string with the appropriate replacement, if any
```

Comment:
```suggestion
 * @param rx: The regular expression to search for
 * @param toSearch: The string to search
 * @param replacement: The string that replaces matches
 * @returns The given string with the appropriate replacements, if any
```

Reply:
Just to be clear, this returns the `toSearch` string with the first match replaced by the replacement?

Reply:
Correct.